Introduce check pointing

davidegrohmann opened the pull request:
Missing: time based check pointing and check pointer running in a separate thread.

Based on top of: https://github.com/neo4j/neo4j/pull/4736

davidegrohmann committed:
Introduce checkpointing as a new strategy for flushing the store

The store is not flushed any more during log rotation but it is
flushed after a configurable number of transactions.  Log pruning has
been moved from log rotation to check poiting in order to make sure we
never prune logs containing latest check points or commits pointed by
the latest wcheck point.

Missing: check pointing based on time and split check pointing in a
different thread

davidegrohmann committed:
Move check pointing in its own thread

Check pointing can be triggered based on time and number of
transactions, default values are 5 minutes and 100_000 txs
respectively.

Introduce CheckPointTracer in order to trace check point events, which
are now disconnected from LogAppendEvents.

tinwelint commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -142,6 +142,26 @@
+                  "check-point interval can also reduce the I/O load that the database places on the system, as each " +
+                  "check-point implies a flushing and forcing of all the store files. The default is '5m' for a " +
+                  "check-point every 5 minutes. Other supported units are 's' for seconds, and 'ms' for milliseconds." )
+    public static final Setting<Long> check_point_max_time_interval = setting( "dbms.checkpoint.time.interval", DURATION, "5m" );
```
Inconsistent naming of these two settings, also to me it makes more sense to name not have `interval` last. My suggestion: `dbms.checkpoint.interval.tx` and `dbms.checkpoint.interval.time`

Also the descriptions are confusing talking about e.g. "maximal time", which to me sounds like the configuration is about specifying the maximum time between two checkpoints, which means that the db can decide to introduce _more_ frequent checkpoints than that.

davidegrohmann replied:
You are right the "maximal time" is confusing. Perhaps we can simply call it "interval time", indeed we can potentially check pointing more frequently than 5 mins (by shutting down the db or by committing many transactions in a short period of time) or even check pointing less frequently than 5 mins (if the store flushing is very slow and takes more than 5 mins).

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointer.java:
```diff
@@ -0,0 +1,43 @@
+
+import java.io.IOException;
+
+public interface CheckPointer
```
I'd like some class-level javadocs here.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/entry/LogEntryWriterV1.java:
```diff
@@ -86,11 +86,10 @@ public void serialize( TransactionRepresentation tx ) throws IOException
         writeLogEntryHeader( CHECK_POINT );
-        channel.putLong( transactionId ).
-                putLong( logPosition.getLogVersion() ).
+        channel.putLong( logPosition.getLogVersion() ).
```
Btw why do we store the log position in the checkpoint record? Just transaction id should be enough.

davidegrohmann replied:
The idea is to save in the check point the log position where to start recovery from. Two benefits:
1) when you find the latest check point entry you know where to jump directly without searching for the commit entry
2) having the log version of the transaction referred by the latest check point helps in making sure that we prune the logs up to that version - 1

Note that potentially you might have the check point and the transaction in different log versions.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/recovery/StoreRecoverer.java:
```diff
@@ -71,21 +75,24 @@ public boolean recoveryNeededAt( File dataDir, long currentLogVersion )
+        NeoStoreUtil neoStoreUtil = new NeoStoreUtil( dataDir, fs );
+        LogPosition lastTxCommittedLogPosition = new LogPosition(
+                neoStoreUtil.getLastCommittedTxLogVersion(), neoStoreUtil.getLastCommittedTxLogByteOffset() );
+        return !checkPoint.getLogPosition().equals( lastTxCommittedLogPosition );
```
To reiterate: I don't see a need for storing the log position in the checkpoint record. If we just had the transaction id we could compare instead of getting the last log position, which feels a bit awkward.

davidegrohmann replied:
See other comment.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/NeoStore.java:
```diff
@@ -147,22 +151,24 @@ public static boolean isStorePresent( FileSystemAbstraction fs, File storeDir )
-    private final OutOfOrderSequence lastCommittedTx = new ArrayQueueOutOfOrderSequence( -1, 200 );
-    private final OutOfOrderSequence lastClosedTx = new ArrayQueueOutOfOrderSequence( -1, 200 );
+    private final OutOfOrderSequence lastCommittedTx = new ArrayQueueOutOfOrderSequence( -1, 200, new long[3] );
+    private final OutOfOrderSequence lastClosedTx = new ArrayQueueOutOfOrderSequence( -1, 200, ZERO );
```
Feels very unnecessary and potentially dangerous going forward to use share that `long[1]` since it's non-zero length.

davidegrohmann replied:
You are right that's dangerous. I'll fix it.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointThreshold.java:
```diff
@@ -0,0 +1,42 @@
+ */
+package org.neo4j.kernel.impl.transaction.log.checkpoint;
+
+public interface CheckPointThreshold
```
Class-level javadocs would be nice.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointerImpl.java:
```diff
@@ -0,0 +1,180 @@
+import static org.neo4j.kernel.impl.transaction.log.checkpoint.CheckPointer.PrintFormat.prefix;
+import static org.neo4j.kernel.impl.util.JobScheduler.Groups.checkPoint;
+
+public class CheckPointerImpl extends LifecycleAdapter implements CheckPointer, Listener<Void>
```
Class-level javadocs would be nice.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CountCommittedTransactionThreshold.java:
```diff
@@ -0,0 +1,64 @@
+
+import org.neo4j.kernel.impl.util.Listener;
+
+public class CountCommittedTransactionThreshold implements CheckPointThreshold, TransactionCommitWatcher
```
Wouldn't it be cheaper to not count transactions, but instead have `checkPointHappened()` accept a `transactionId`, keep that in a field here and just compare with that value in `newTransactionCommitted()`? I'm thinking that this may be an overhead since we scale well with concurrent committers and it feels unnecessary to introduce contention here.

davidegrohmann replied:
Absolutely! That's a great idea!

tinwelint replied:
That field doesn't even need to be volatile, I mean practically it will work and only potentially have the discovery of need for check point slightly delayed, which is absolutely fine. Just document that fact if you do that.

chrisvest replied:
Benign data races are hard to reason about, so don't add them unless a compelling performance boost can be demonstrated.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/TransactionCommitWatcher.java:
```diff
@@ -0,0 +1,34 @@
+ */
+package org.neo4j.kernel.impl.transaction.log.checkpoint;
+
+public interface TransactionCommitWatcher
```
Class-level javadocs would be nice.

davidegrohmann commented:
@tinwelint I have addressed your comments plus making sure we stop the check pointing service cleanly on shutdown. Mind to check again if everything is in order?

tinwelint commented:
There seems to be cypher scala compilation errors

chrisvest commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -142,6 +142,27 @@
+                  "store files. The default is '5m' for a check-point every 5 minutes. Other supported units are 's' " +
+                  "for seconds, and 'ms' for milliseconds." )
+    public static final Setting<Long> check_point_interval_time = setting( "dbms.checkpoint.interval.time", DURATION, "5m" );
+
```
I wonder if IO rate is a better way of holding back, such that the check-pointer runs continuously, but only eat up a limited number of IOPS to avoid starving more time critical work, like page faulting and writing to the log.

davidegrohmann replied:
I am not sure I understand what you are suggesting here.

chrisvest replied:
I am suggesting something along the lines of https://github.com/neo4j/neo4j/blob/2.3/community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCache.java#L855-L969 but targeting a specific IOPS rate instead of some time percentage.

davidegrohmann replied:
That sounds interesting. Perhaps better if we try that in a later PR?

chrisvest replied:
Sure.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -1123,6 +1142,14 @@ public synchronized void stop()
+    {
+        while ( !neoStoreModule.neoStore().closedTransactionIdIsOnParWithOpenedTransactionId() )
+        {
+            LockSupport.parkNanos( 1_000_000 ); // 1 ms
```
Waiting less than 10 milliseconds will change the system-wide process scheduling quantum on Windows.

davidegrohmann replied:
I have simply moved that method from 'LogRotationControl'. We have that today in the code: https://github.com/neo4j/neo4j/blob/2.3/community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/rotation/LogRotationControl.java#L50

I am not against fixing it, but doing it in this PR feels wrong.

chrisvest replied:
Fair enough.

davidegrohmann commented:
@tinwelint I'll rebase again, hopefully they will go away

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/PartialTransactionFailureIT.java:
```diff
@@ -184,6 +185,8 @@ public void run()
                     latch.await();
                     db.getDependencyResolver().resolveDependency( LogRotation.class ).rotateLogFile();
+                    db.getDependencyResolver().resolveDependency( CheckPointer.class ).forceCheckPoint();
+                    tx.close();
```
Explicitly closing the transaction should not be necessary.

davidegrohmann replied:
Right. I messed up the merge, I'll clean up.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/NeoStore.java:
```diff
@@ -893,14 +907,18 @@ public long nextCommittingTransactionId()
+            setRecord( Position.LAST_TRANSACTION_LOG_BYTE_OFFSET, transactionData[3] );
             lastTransactionChecksum = checksum;
+            lastTransactionLogVersion = logVersion;
+            lastTransactionLogByteOffset = byteOffset;
```
The boolean returned by `lastCommittedTx.offer()` means that only the transaction with the greatest transactionId will have its logVersion and byteOffset (along with the other data) published? Which should be fine because the transaction id is representative of the order of the transactions in the log because the BatchingTransactionAppender both produces the transaction id and appends the transaction to the log, in the same critical section locking the log file.

Yeah... I think this is good.

davidegrohmann replied:
:+1:

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointerImpl.java:
```diff
@@ -0,0 +1,158 @@
+                        }
+                    }
+                }
+            }
```
This is not good. When you synchronise on the logFile, you also prevent transactions from committing. You should only lock on it for writing out the check-point entries, and perhaps for reading the LogPosition of the last commit. Use a different strategy for ensuring that only one check-point is in progress at a time. You could for instance have an AtomicBoolean, where you only call doCheckPoint if you can successfully compareAndSet it to true, for instance.

davidegrohmann replied:
Very good point. The synchronized should be moved only around the writing of the check pointing in the transaction log. We don't even need the `AtomicBoolean` since there are never 2 running check point threads at the same time.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointerImpl.java:
```diff
@@ -0,0 +1,158 @@
+            transactionLogWriter.checkPoint( logPosition );
+            channel.emptyBufferIntoChannelAndClearIt();
+            channel.force();
+        }
```
I wonder if you could get the BatchingTransactionAppender to write the check-point entry for you, so you go through the write-path where concurrent `channel.force()` calls gets coalesced.

chrisvest replied:
Good comments in this class, though.

davidegrohmann replied:
I'll look into that, sounds very promising. Can we handle that in a separate PR? or do you think it must be part of this one?

chrisvest replied:
I think it can be separate.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointer.java:
```diff
@@ -0,0 +1,66 @@
+        public void forceCheckPoint()
+        {
+        }
+    };
```
This implementation isn't used anywhere. And it's also kind of dangerous.

davidegrohmann replied:
I'll delete it.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointScheduler.java:
```diff
@@ -0,0 +1,83 @@
+        }
+    };
+
+    private boolean stopped = false;
```
This field should be volatile

davidegrohmann replied:
Don't see what that change will improve.

chrisvest replied:
It will ensure that the write in `stop()` is visible to this read in, which happens in a different thread than the write.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointScheduler.java:
```diff
@@ -0,0 +1,83 @@
+            }
+            catch ( IOException e )
+            {
+                throw new UnderlyingStorageException( e );
```
If smth happens than checkpointing thread will simply die silently, not reschedule the job and we won't write any more checkpoint log entries. This looks harsh. Maybe we can do rescheduling based on kernel health? What if checkpointing thread met some transient IO error?

davidegrohmann replied:
I will add a `kernel.panic()` in `CheckPointerImpl` in case it fails to write the check point in the log like: https://github.com/neo4j/neo4j/blob/2.3/community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/BatchingTransactionAppender.java#L241. Don't see how rescheduling will help here.

MishaDemianenko replied:
will agree with @lutovich here: in case if we will encounter any problem during `checkPointIfNeeded` invocation we will throw Exception and will never reschedule itself. Is it what we really want here?

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointerImpl.java:
```diff
@@ -0,0 +1,143 @@
+
+        synchronized ( logFile )
+        {
+            synchronized ( channel ) // TODO: is this needed?
```
is it?

tinwelint replied:
This code is simply moved from somewhere else (I think I added it), and I _do_ think it's needed. Anyway that could be the topic of another, future, revisit.

davidegrohmann replied:
I agree. It is needed. I'll remove the TODO.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/util/Neo4jJobScheduler.java:
```diff
@@ -142,6 +154,7 @@ public void shutdown()
             if(scheduledExecutor != null)
             {
                 scheduledExecutor.shutdown();
+                System.out.println( scheduledExecutor.getQueue() );
```
Line could be removed

lutovich commented on community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointerIntegrationTest.java:
```diff
@@ -0,0 +1,224 @@
+        }
+
+        // trigger the check point
+        Thread.sleep( TimeUnit.SECONDS.toMillis( seconds ) );
```
_Not really important:_ line could be written like this `TimeUnit.SECONDS.sleep(seconds)`

chrisvest replied:
Could alternatively use a monitor or tracer to wait for the check point, and then check if it happened within a reasonable amount of time (with plenty of slack, so it won't be flaky).

davidegrohmann replied:
I would have gone with a tracer but they are not easily configurable and definitely not via graph database configuration. I am unsure if the extra complexity will pay off, since this test never failed any PR build so far.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointScheduler.java:
```diff
@@ -0,0 +1,83 @@
+            if ( !stopped )
+            {
+                handle = scheduler.schedule( checkPoint, job, recurringPeriodMillis, MILLISECONDS );
+            }
```
Why don't you use `scheduleRecurring()`? That way you won't have to worry about stray exceptions killing the job.

davidegrohmann replied:
if you do that, than you have to deal with the case when you schedule check points faster than writing them

davidegrohmann committed:
Check pointing and log rotation stress test

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreMigrator.java:
```diff
@@ -226,6 +235,14 @@ private void writeLastTxChecksum( File migrationDir, long lastTxChecksum ) throw
 
+    private void writeLastTxLogPosition( File migrationDir, LogPosition lastTxLogPosition ) throws IOException
+    {
+        try ( Writer writer = fileSystem.openAsWriter( lastTxLogPositionFile( migrationDir ), "utf-8", false ) )
```
Charsets.UTF_8.name() instead of `utf-8`?

davidegrohmann committed:
CheckPointer will use TransactionAppender for writing check points in the log

In this way the log force needed by the check pointing can be batched together
with several transaction commits.  This should improve performance and reduce
contantion around log force.

tinwelint commented on stresstests/src/test/java/org/neo4j/kernel/stresstests/CheckPointingLogRotationStressTesting.java:
```diff
@@ -0,0 +1,112 @@
+        String pageCacheMemory = fromEnv( "CHECK_POINT_LOG_ROTATION_PAGE_CACHE_MEMORY", DEFAULT_PAGE_CACHE_MEMORY );
+        String pageSize = fromEnv( "CHECK_POINT_LOG_ROTATION_PAGE_SIZE", DEFAULT_PAGE_SIZE );
+
+        System.out.println( "1/5  Building initial store..." );
```
We probably don't want these `System.out.println` here

davidegrohmann replied:
It is a stress test that will run separately from the main build so I think it is fine in having some printing. We can reconsider if this causes problems.
